src/Microsoft.Health.Fhir.Azure/KeyVault/KeyVaultSecretStore.cs

```diff
@@ -21,20 +25,67 @@ public class KeyVaultSecretStore : ISecretStore
         private IKeyVaultClient _keyVaultClient;
         private Uri _keyVaultUri;
 
+        private const int DefaultRetryCount = 3;
+        private static readonly Func<int, TimeSpan> DefaultSleepDurationProvider = new Func<int, TimeSpan>(retryCount => TimeSpan.FromSeconds(Math.Pow(2, retryCount)));
+        private static readonly HttpStatusCode[] _defaultStatusCodesToRetry =
+        {
+            HttpStatusCode.RequestTimeout,
+            HttpStatusCode.ServiceUnavailable,
+        };
+
+        private readonly AsyncRetryPolicy _retryPolicy;
+
         public KeyVaultSecretStore(IKeyVaultClient keyVaultClient, Uri keyVaultUri)
+            : this(keyVaultClient, keyVaultUri, DefaultRetryCount, DefaultSleepDurationProvider, _defaultStatusCodesToRetry)
+        {
+        }
+
+        public KeyVaultSecretStore(IKeyVaultClient keyVaultClient, Uri keyVaultUri, int retryCount, Func<int, TimeSpan> sleepDurationProvider)
+            : this(keyVaultClient, keyVaultUri, retryCount, sleepDurationProvider, _defaultStatusCodesToRetry)
+        {
+        }
+
+        public KeyVaultSecretStore(
+            IKeyVaultClient keyVaultClient,
+            Uri keyVaultUri,
+            int retryCount,
+            Func<int, TimeSpan> sleepDurationProvider,
+            HttpStatusCode[] statusCodesToRetry)
         {
             EnsureArg.IsNotNull(keyVaultClient, nameof(keyVaultClient));
             EnsureArg.IsNotNull(keyVaultUri, nameof(keyVaultUri));
+            EnsureArg.IsGte(retryCount, 0, nameof(retryCount));
+            EnsureArg.IsNotNull(sleepDurationProvider, nameof(sleepDurationProvider));
+            EnsureArg.IsNotNull(statusCodesToRetry, nameof(statusCodesToRetry));
 
             _keyVaultClient = keyVaultClient;
             _keyVaultUri = keyVaultUri;
+
+            _retryPolicy = Policy.Handle<KeyVaultErrorException>(kve => statusCodesToRetry.Contains(kve.Response.StatusCode))
+                .WaitAndRetryAsync(retryCount, sleepDurationProvider);
         }
 
         public async Task<SecretWrapper> GetSecretAsync(string secretName, CancellationToken cancellationToken)
         {
             EnsureArg.IsNotNullOrWhiteSpace(secretName);
 
-            SecretBundle result = await _keyVaultClient.GetSecretAsync(_keyVaultUri.AbsoluteUri, secretName, cancellationToken);
+            SecretBundle result;
+            try
+            {
+                result = await _retryPolicy.ExecuteAsync(async () =>
```

Comment:
nit: ExecuteAsync takes Func<Task> so this could just be simplified as `await _retryPolicy.ExecuteAsync(() => _keyVaultClient.GetSecretAsync(...))`.